### PR TITLE
Added cache_disk

### PR DESCRIPTION
**Pull request by dploeger**
## Summary
This deprecates apache::mod::disk_cache and instead adds support for apache::mod::cache_disk. Also adds parameters to both apache::mod::cache_disk and apache::mod::cache.

## Additional Context
Since Apache 2.4 mod_disk_cache is named mod_cache_disk. This officially supports mod_cache_disk and additionally adds configuration options for it. To be backwards compatible, the old module is still available and simply uses the new module.

## Checklist
- [x] 🟢 Spec tests.
- [x] 🟢 Acceptance tests.
- [x] Manually verified. (For example `puppet apply`)

### CODE REVIEW

File: manifests/mod/cache_disk.pp
Comment (ekohl) [inline manifests/mod/cache_disk.pp:116]: Please use EPP for new templates.

File: manifests/mod/cache_disk.pp
Comment (ekohl) [inline manifests/mod/cache_disk.pp:77]: Please use `$facts['os']['family']`

**Review by ekohl — CHANGES_REQUESTED** [review]
We've dropped support for Apache < 2.4 so you can simplify a few things.

File: manifests/mod/cache_disk.pp
Comment (dploeger) [reply manifests/mod/cache_disk.pp:77]: Oh, you're quick. 😆  The PR is in draft. I'm currently optimizing and adding tests. Thanks!

File: manifests/mod/cache_disk.pp
Comment (dploeger) [reply manifests/mod/cache_disk.pp:116]: Oh, okay. I'll change that.

File: manifests/mod/cache_disk.pp
Comment (ekohl) [reply manifests/mod/cache_disk.pp:77]: I saw that and didn't look too closely at details, just wanted to point out some general things that jumped out and can affect writing tests.

File: manifests/mod/cache.pp
Comment (smortex) [inline manifests/mod/cache.pp:28]: This allow multiple values, we should use:

```suggestion
  Array[String[1]] $cache_ignore_headers                               = [],
```

File: manifests/mod/cache_disk.pp
Comment (smortex) [inline manifests/mod/cache_disk.pp:40]: Are relative path acceptable?  If yes, a non-null `String` should be expected:

```suggestion
  Optional[String[1]] $cache_root              = undef,
```

otherwise:

```suggestion
  Optional[Stdlib::Absolutepath] $cache_root              = undef,
```

**Review by smortex — COMMENTED** [review]
There are a few minor things that we can improve.  I added in-line comments. Thanks!

**Comment by dploeger** [comment]
@smortex  Thanks. Done.

File: manifests/mod/disk_cache.pp
Comment (smortex) [inline manifests/mod/disk_cache.pp:34]: For backwards compatibility we may need something like:

```suggestion
    cache_ignore_headers    => $cache_ignore_headers.split(' '),
```

(I don't think it's worth breaking backwards compatibility to improve a legacy interface)

File: spec/classes/mod/cache_disk_spec.rb
Comment (smortex) [inline spec/classes/mod/cache_disk_spec.rb:32]: Not sure about the suggestion from rubocop :upside_down_face: 

The line length feel reasonable with current editors and screen so would not mind seeing everything on the same line:

```suggestion
    it { is_expected.to contain_file('cache_disk.conf').with(content: default_config) }
```


Note that I replaced `expect(subject)` by `is_expected`: it does the same but makes the whole test [almost] readable as a sentence.

**Comment by dploeger** [comment]
@smortex Should I take a look at the other Rubocop findings?

**Comment by smortex** [comment]
> @smortex Should I take a look at the other Rubocop findings?

Yes please :smile:  They are admittedly mostly gratuitous, but allow some consistency across the module which makes maintenance easier.  Rubocop has to pass for the unit and integration tests to run.

As it is now, it looks quite good, CI will maybe help us spot some remaining issues.  I am just wondering if the unit tests for the legacy class should be updated to check that it instantiate the new class with the right parameters and not more.

**Comment by dploeger** [comment]
Fixed Rubocop

File: manifests/mod/cache_disk.pp
Comment (ekohl) [inline manifests/mod/cache_disk.pp:14]: You can assume Apache 2.4 only and I don't see Apache 2.2 references in this class

File: manifests/mod/cache_disk.pp
Comment (ekohl) [inline manifests/mod/cache_disk.pp:35]: I don't see the code for Apache 2.2.

File: manifests/mod/cache_disk.pp
Comment (ekohl) [inline manifests/mod/cache_disk.pp:37]: I think this is sufficient
```suggestion
# @see https://httpd.apache.org/docs/2.4/mod/mod_cache_disk.html
```

File: manifests/mod/cache_disk.pp
Comment (ekohl) [inline manifests/mod/cache_disk.pp:68]: Isn't this what `pick()` does?
```suggestion
  $_configuration_file_name = pick($configuration_file_name, 'cache_disk.conf')
```

File: manifests/mod/cache_disk.pp
Comment (dploeger) [reply manifests/mod/cache_disk.pp:35]: Oh right, I missed that comment.

File: manifests/mod/cache_disk.pp
Comment (dploeger) [reply manifests/mod/cache_disk.pp:68]: True, but I wanted to rely on built in features instead of stdlib uses. Should I change it to pick?

Comment (ekohl) [reply manifests/mod/cache_disk.pp:65]: What's one more:
```console
$ rg pick
manifests/custom_config.pp
97:  $_file_mode = pick($file_mode, $apache::file_mode)

manifests/mod/userdir.pp
42:    'userdir'         => pick($userdir, $path),

manifests/mod/mime_magic.pp
13:  $_magic_file = pick($magic_file, "${apache::conf_dir}/magic")

manifests/mod/mime.pp
21:  $_mime_types_additional = pick($mime_types_additional, $apache::mime_types_additional)
```

Comment (dploeger) [reply manifests/mod/cache_disk.pp]: Okay. 🤷🏽

File: templates/mod/cache.conf.epp
Comment (smortex) [inline templates/mod/cache.conf.epp:29]: In this case (list of words), prefer `join` which avoid a trailing space after the last item and is a bit more readable:

```suggestion
CacheIgnoreHeaders <% $cache_ignore_headers.join(' ') %>
```

Comment (ThomasMinor) [reply templates/mod/cache.conf.epp]: Nice. Added a missing '='.

File: manifests/mod/disk_cache.pp
Comment (smortex) [inline manifests/mod/disk_cache.pp:34]: This cause trouble if `$cache_ignore_headers` is `undef`.  We can default to something like:

```suggestion
    cache_ignore_headers    => if $cache_ignore_headers {
                                 $cache_ignore_headers.split(' ')
                               } else {
                                 []
                               },
```

**Review by smortex — COMMENTED** [review]
It also look like the `apache::mod::disk_cache` is included in `manifests/default_mods.pp` for FreeBSD.

I just checked on my FreeBSD box, and it seems the default config does not include the `cache_disk` module, so I think we can also remove these lines:

https://github.com/puppetlabs/puppetlabs-apache/blob/9f159df9c66807168c3ff495be2b5aeac4319b9b/manifests/default_mods.pp#L72-L73

(In fact, all these lists seems outdated, maybe some global rework would be great, but that is out-of-scope of this PR)

File: manifests/mod/disk_cache.pp
Comment (ThomasMinor) [reply manifests/mod/disk_cache.pp:34]: I actually prefer the selector variant here.

Comment (ThomasMinor) [reply manifests/mod/disk_cache.pp]: But actually the splitting is wrong here. I'll asap commit a fixed version.

**Comment by ThomasMinor** [comment]
It seems the failures during acceptance testing do not relate to the changes in this PR. Any idea when these might get fixed? Can I offer some support here?

**Comment by smortex** [comment]
Puppet 8 seems to add a check for duplicates that fail.  I opened #2525 to address the issue.

**Comment by smortex** [comment]
@dploeger can you please rebase your changes on top of the main branch?  This should fix CI.

From your working directory:

```
git fetch origin       # Download the latest code we have here
git rebase origin/main # Move your commits on top of the main branch
git push -f            # Send the changes (-f is required because we re-wrote history)
```

**Comment by ThomasMinor** [comment]
Hi @smortex, I just merged the latest changes from main to this repo. Since the 2 failing tests do not seem to be related to the changes in this PR, how do we proceed from here?

**Comment by smortex** [comment]
@ThomasMinor From my point of view this look okay.  The repository require multiple approval before merging so the next move is probably to have some feedback from @puppetlabs maybe?

I re-approved the CI run (any change you do to your commit invalidate the test), I expect it to all pass except the 2 known-bad tests which I think we can ignore.

**Comment by ThomasMinor** [comment]
Hi @smortex, it seems another test joined the club and failed with connection failures. Can I assist solving the problem?

**Comment by ThomasMinor** [comment]
Seems I looked wrong, 2 fail, one test was skipped.

**Comment by smortex** [comment]
@ThomasMinor the Ubuntu ARM tests are broken and mend seems to only run when the PR is from a branch in the repo itself.

`spec/classes/mod/disk_cache_spec.rb` is maybe testing too much things given the corresponding code is only a wrapper around `apache::mod::cache_disk` (it act as an integration test rather than a unit test).

Also there is a merge commit in the branch which is not nice. Rebasing on top of master would fix this.

These two annoyances are not stoppers for me.

**Comment by ThomasMinor** [comment]
How about a final review, @ekohl & @bastelfreak?
Then we might be able to close this one.

**Review by ekohl — APPROVED** [review]

**Pull request merged**
